### PR TITLE
Impl HasSynced for memory controller and improve HasSynced of ADSC

### PR DESCRIPTION
**Pull request by YonkaFang**
**Please provide a description of this PR:**

In current istio codebase, memory config controller does not impl the `HasSynced` method which means we will wait for readiness of config source using mem store (such as mcp-over-xds). And this will break the integrity of istiod state control.

This PR impl `HasSynced` for mem controller and use it in the XDS configsource case.

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [X] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [X] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Review by hzxuzhonghu — COMMENTED** [review]
Can you add some tests

**Comment by YonkaFang** [comment]
> Can you add some tests

ok. i'll push them a little later

**Comment by YonkaFang** [comment]
@hzxuzhonghu I've pushed the test cases. Plz help take another look.

File: pilot/pkg/bootstrap/configcontroller.go [hunk 1]
```diff
@@ -221,7 +221,9 @@ func (s *Server) initConfigSources(args *PilotArgs) (err error) {
 				return fmt.Errorf("failed to dial XDS %s %v", configSource.Address, err)
 			}
 			store := memory.MakeSkipValidation(collections.Pilot)
-			configController := memory.NewController(store)
+			configController := memory.NewControllerWithOptions(store, &memory.Config{
+				HasSynced: xdsMCP.HasSynced,
```
Comment (hzxuzhonghu) [inline pilot/pkg/bootstrap/configcontroller.go:225]: why not directly pass `func HasSynced() bool `

File: pilot/pkg/config/memory/controller.go [hunk 2]
```diff
@@ -26,9 +26,35 @@ import (
 	"istio.io/istio/pkg/config/schema/collection"
 )
 
+type Config struct {
+	HasSynced func() bool
+	SyncProc  bool
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/memory/controller.go:31]: what's this?

Comment (YonkaFang) [reply pilot/pkg/config/memory/controller.go:31]: It keeps the ability to new a controller with sync monitor. U can see the code below.

File: pilot/pkg/config/memory/controller.go [hunk 3]
```diff
@@ -26,9 +26,35 @@ import (
...
+	if opts == nil {
+		opts = &Config{}
+	}
+
+	out := &controller{
+		configStore: cs,
+		opts:        opts,
+	}
+
+	if opts.SyncProc {
```
Comment (hzxuzhonghu) [inline pilot/pkg/config/memory/controller.go:51]: It looks no better than before

**Review by hzxuzhonghu — COMMENTED** [review]
Basically lg, but need some improve

File: pilot/pkg/bootstrap/configcontroller.go (hunk 1 above)
Comment (YonkaFang) [reply pilot/pkg/bootstrap/configcontroller.go:225]: I consider the fact that there're several `new` functions already. Then I add a general one which can do the complex init things.

File: pilot/pkg/config/memory/controller.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/config/memory/controller.go:31]: Yes, understood. Looks below is more readable

func NewController(cs model.ConfigStore, hasSynced func() bool) model.ConfigStoreCache {

File: pilot/pkg/config/memory/controller.go (hunk 3 above)
Comment (YonkaFang) [reply pilot/pkg/config/memory/controller.go:51]: just keep a compability with existing usage.

File: pilot/pkg/config/memory/controller.go (hunk 2 above)
Comment (YonkaFang) [reply pilot/pkg/config/memory/controller.go:31]: I think the `hasSynced` func is not a common args. It's more like a customization. So I didn't add it to the **common** `NewController` func.
BTW, there've been tons of usage of `NewController`.

**Comment by YonkaFang** [comment]
@hzxuzhonghu Plz help review the updated code. Thanks.

**Review by hzxuzhonghu — APPROVED** [review]
Nit: The patch function is still very complex, i will clean up afterwards.

**Pull request merged**